homework/zad9_opencv/line_match2.cpp

```diff
@@ -0,0 +1,161 @@
+#include <opencv2/core.hpp>
+#include <opencv2/imgproc.hpp>
+#include <opencv2/highgui.hpp>
+
+#include <iostream>
+#include <mutex>
+#include <future>
+#include <thread>
+
+void show_wait_destroy(const char* winname, cv::Mat img) {
+    cv::imshow(winname, img);
+    cv::moveWindow(winname, 500, 0);
+    cv::waitKey(0);
+    cv::destroyWindow(winname);
+}
+
+cv::Mat createMatLineAlpha( const cv::Size& size, int iAlpha )
+{
+    cv::Mat mat(size, CV_8U);
+    cv::Point origin( size.width / 2, size.height / 2 );
+    double dRad = ( iAlpha * M_PI ) / 180.0;
+
+    if ( iAlpha != 90 )
+    {
+        for( int iCol = 0; iCol < size.width; ++iCol )
+        {
+            double dValue = std::tan(dRad) * ( iCol - origin.x );
+            int iPos = origin.y - std::lround(dValue);
+
+            if ( iPos >= 0 && iPos < size.height )
+            {
+                mat.at<unsigned char>( iPos , iCol) = 1;
+            }
+        }
+    }
+    else
+    {
+        for ( int iRow = 0; iRow < size.height; ++iRow )
+        {
+            mat.at<unsigned char>(iRow, origin.x) = 1;
+        }
+    }
+
+    return mat;
```

Comment:
taki pomysł podrzucam - a jakby stworzyć cv::Mat z linią poziomą i użyć gotowej funkcji rotate? Chyba taka jest openCV. Myślę, że mniej kodu by tutaj wyszło.

Reply:
Nie do konca. Zrobilem test takiej funkcji w test_ocv.cpp.
Linia o dlugosci np 7px po obrocie o 45 stopni nie wypełni narożników.

Przykład:
[ 0] > 0 0 0 0 0 0 0 
[ 1] > 0 0 0 0 0 0 0 
[ 2] > 0 0 0 0 0 0 0 
[ 3] > 1 1 1 1 1 1 1 
[ 4] > 0 0 0 0 0 0 0 
[ 5] > 0 0 0 0 0 0 0 
[ 6] > 0 0 0 0 0 0 0 
Po obrocie o 45 stopni
[ 0] > 0 0 0 0 0 0 0 
[ 1] > 0 0 0 0 0 1 0 
[ 2] > 0 0 0 0 1 0 0 
[ 3] > 0 0 0 1 0 0 0 
[ 4] > 0 0 1 0 0 0 0 
[ 5] > 0 1 0 0 0 0 0 
[ 6] > 0 0 0 0 0 0 0

Zobacz rownież  przykład:
http://opencvexamples.blogspot.com/2014/01/rotate-image.html
Tutaj tez widać, że po obrocie o kat np 10 stopni pozostaja czarne piksele co jest logiczne.

"Moja" fukcja dba rownież o narożniki.

Reply:
Masz rację, tak by to wyglądało. A w takim razie, gdyby utworzyć macierz z linią po przekątnej (funkcja `eye()`) to już by nie było tego problemu.